openssl/src/base64.rs

```diff
@@ -0,0 +1,170 @@
+use std::ptr;
+use std::slice;
+
+use ffi;
+use error::ErrorStack;
+use {cvt_p, cvt};
+
+struct Bio(*mut ffi::BIO);
+
+impl Bio {
+    pub fn as_ptr(&self) -> *mut ffi::BIO {
+        self.0
+    }
+}
+
+impl Drop for Bio {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::BIO_free_all(self.0);
+        }
+    }
+}
+
+/// Encode `data` in base64. If `newlines` is true then the data is
+/// encoded with line breaks.
+pub fn encode(data: &[u8], newlines: bool) -> Result<Vec<u8>, ErrorStack> {
+    if data.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let bio = unsafe { try!(cvt_p(ffi::BIO_new(ffi::BIO_f_base64()))) };
+
+    let bio = Bio(bio);
+
+    if !newlines {
+        unsafe {
+            ffi::BIO_set_flags(bio.as_ptr(), ffi::BIO_FLAGS_BASE64_NO_NL);
+        }
+    }
+
+    unsafe {
+        // Add some memory to hold the result
+        let mem = try!(cvt_p(ffi::BIO_new(ffi::BIO_s_mem())));
+
+        try!(cvt_p(ffi::BIO_push(bio.as_ptr(), mem)));
+
+        // Write `data` to the BIO
+        try!(cvt(ffi::BIO_write(bio.as_ptr(),
+                                data.as_ptr() as *const _,
+                                data.len() as _)));
```

Comment:
I'll admit that I was a bit optimistic here. What's the proper way to handle short read/writes? Can I just return `ErrorStack::get()` in these situations?